### ISSUE TITLE
Is there any support (or planned support) for failover?

### ISSUE THREAD

**Issue by CorwinTanner** [CopernicaMarketingSoftware/AMQP-CPP#327]
Many AMQP libraries accept a list of connections to handle graceful failover in certain redundancy scenarios, but I see no such functionality in AMQP-CPP (unless it's under the hood during connection negotiations). One could build a failover layer on top, but without direct access to the pending message queue there is a potential for loss of data.

Is there any current or planned support for this functionality?

Thank you for your work on this library.

**Comment by EmielBruijntjes** [comment]
This is not at all trivial. If the connection to RabbitMQ suddenly breaks, there are a lot of choices to be made. For example:

- The messages that have been sent, but for which no confirmation has yet been received, should we consider them to be received by RabbitMQ, or should they be retried once we've set up a new connection?
- Messages that have recently been published (publish operations are normally not confirmed, it's a one-way-message), should we assume that such messages have been received and handled (possibly stored) by RabbitMQ, or should we assume that such messages are lost and should we retry them? And how many seconds should we go back in time for retrying? Everything that we sent during the last 0.1 second? Last full second? 
- If we have still a queue of messages when the connection is lost, should we send that same queue after we've established a new connection? Messages in the queue might be dependent on a message that was sent earlier (for example the instruction to publish a message might only be meaningful if it was sent after the instruction to create and bind the exchange. If we just send the publish-instruction to the next server, it will get lost as it will not reach the right exchange or queue).
- Consumed messages can no longer be acknowledged. Should the AMQP library withdraw such messages via a special call to user space? What should the library do if they are acknowledged? What should the AMQP library do if we start consuming on the new connection, and the same delivery-tags are handed out?

The point is that it very much depends on the nature of the application what you want to do when a connection breaks. For a very simple application that does nothing else than publishing non-critical messages, a simple failover system can work where you start publishing to the next available RabbitMQ server. But for uses that are even a little more advanced, your application probably needs its own handling of a broken connection. That's why we have not implemented such a automatic failover mechanism inside the AMQP library. It is up to the application to implement the ConnectionHandler::onError() method, and do the appropriate handling: for example to inform the user that something totally unexpected has happened, and exit the application. Or try to automatically set up a new connection, run the right instructions to initialize the AMQP environment (for example create the exchanges and queues, start transactions, etc), and continue publishing messages.

If you want to prevent losing data inside AMQP-CPPs internal queue, you should avoid that messages end up in that queue. For every operation you should implement the onSuccess handler, and send the next instruction only after the initial instruction has been completed. In such a setup there is no need for AMQP-CPP to put things in a queue, and everything can immediately be sent out over the network. Although even then, if there is a hardware or software failure, you can never be sure whether your messages has received the RabbitMQ server, and whether RabbitMQ managed to store your message.

In short, to answer your question: no, there are no such plans.
